pyro/infer/reparam/stable.py

```diff
@@ -0,0 +1,198 @@
+    """
+    def __call__(self, name, fn, obs):
+        fn, event_dim = self._unwrap(fn)
+        assert isinstance(fn, dist.Stable)
+        assert obs is None, "StableReparam does not support observe statements"
+
+        # Draw parameter-free noise.
+        proto = fn.stability
+        half_pi = proto.new_full(proto.shape, math.pi / 2)
+        one = proto.new_ones(proto.shape)
+        u = pyro.sample("{}_uniform".format(name),
+                        self._wrap(dist.Uniform(-half_pi, half_pi), event_dim))
+        e = pyro.sample("{}_exponential".format(name),
+                        self._wrap(dist.Exponential(one), event_dim))
+
+        # Differentiably transform.
+        x = _standard_stable(fn.stability, fn.skew, u, e)
+        value = fn.loc + fn.scale * x
+
+        # Simulate a pyro.deterministic() site.
+        new_fn = dist.Delta(value, event_dim=event_dim).mask(False)
+        return new_fn, value
+
+
+class SymmetricStableReparam(Reparam):
+    """
+    Auxiliary variable reparameterizer for symmetric
+    :class:`~pyro.distributions.Stable` random variables (i.e. those for which
+    ``skew=0``).
+
+    This is useful in inference of symmetric
+    :class:`~pyro.distributions.Stable` variables because the
+    :meth:`~pyro.distributions.Stable.log_prob` is not implemented.
+
+    This reparameterizes a symmetric :class:`~pyro.distributions.Stable` random
+    variable as a totally-skewed (``skew=1``)
+    :class:`~pyro.distributions.Stable` scale mixture of
+    :class:`~pyro.distributions.Normal` random variables. See Proposition 3. of
+    [1] (but note we differ since :class:`Stable` uses Nolan's continuous S0
+    parameterization).
+
+    [1] Alvaro Cartea and Sam Howison (2009)
+        "Option Pricing with Levy-Stable Processes"
+        https://people.maths.ox.ac.uk/~howison/papers/levy.pdf
+    """
+    def __call__(self, name, fn, obs):
+        fn, event_dim = self._unwrap(fn)
+        assert isinstance(fn, dist.Stable)
+        if is_validation_enabled():
+            if not (fn.skew == 0).all():
+                raise ValueError("SymmetricStableReparam found nonzero skew")
+            if not (fn.stability < 2).all():
+                raise ValueError("SymmetricStableReparam found stability >= 2")
+
+        # Draw parameter-free noise.
+        proto = fn.stability
+        half_pi = proto.new_full(proto.shape, math.pi / 2)
+        one = proto.new_ones(proto.shape)
+        u = pyro.sample("{}_uniform".format(name),
+                        self._wrap(dist.Uniform(-half_pi, half_pi), event_dim))
+        e = pyro.sample("{}_exponential".format(name),
+                        self._wrap(dist.Exponential(one), event_dim))
+
+        # Differentiably transform to scale drawn from a totally-skewed stable variable.
+        _, z = _unsafe_standard_stable(fn.stability / 2, 1, u, e)
+        assert (z >= 0).all()
+        scale = fn.scale * z * (math.pi / 4 * fn.stability).cos().pow(2 / fn.stability)
```

Comment:
I think we should use `( 2 * z) ** 0.5` instead of `z` here (if z is a mixing variable). Btw, what is the role of `(math.pi / 4 * fn.stability).cos().pow(2 / fn.stability)`? I can't find relevant information in Proposition 3 of reference [1] (probably you used a different reference?).

Reply:
Oops sorry wrong link 😊 I've fixed, [here it is](https://people.maths.ox.ac.uk/~howison/papers/levy.pdf
). Note that reference uses a different parameter ordering: `S_stability(scale, skew, loc)`. Let me know if you find a clearer source of this theorem. I really enjoyed that paper, it is very clear.

Reply:
:D In that paper, the mixing "scale" (not variance) will be `z ** 0.5`. In addition, Stable(2, 0, k, ...) is Normal(..., sqrt(2) * k), so as mentioned in my last comment, we should use `( 2 * z) ** 0.5` instead of `z` of `z` here.

Similarly, for the latter term, we should use `pow(1 / fn.stability)` instead of `pow(2 / fn.stability)`.

Reply:
Thanks, I have updated to use `z.sqrt()`, but there is still a bug. Can you explain your `.pow(1/stability)` suggestion? I believe `.pow(2/stability)` is the correct translation of `cos(...).pow(alpha' / alpha)`.

Reply:
I think `.pow(2 / stability)` is the scale of `Y` in the above reference, and we have Z = Y^1/2 X. So we need to multiply `scale` with `...pow(1/stability)`

Reply:
Re bug: If there is still a bug, could you try to completely remove the latter term? In [this reference](https://arxiv.org/pdf/1604.07487.pdf) (at the end of page 8), IIUC there is no such latter term (though I think that your reference is correct).

Reply:
Thanks for explaining, that seems to fix the test!